transformer_engine/pytorch/attention.py

```diff
@@ -173,7 +173,7 @@ def forward(
             output_size[0] * output_size[1],
             output_size[2],
             output_size[3],
-            dtype=query_layer.dtype,
+            dtype=torch.float32,
```

Comment:
What is this for?

Reply:
This is to address the 2nd issue in the description. torch.empty translates to a set of ONNX nodes in the final graph. One of the nodes is a ConstantOfShape node. Since the `dtype` in the original code is set to BF16 (for a BF16 model), this causes the `value` attribute of ConstantOfShape node to be BF16. The [ONNX Spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#ConstantOfShape) doesn't support that type yet and we run into a type error during ONNX Runtime Verification. 
Also open to any other solutions to address this problem

Reply:
The current issue is that it would be a bug for the non onnx cases. Let's use the suggestion below?

Reply:
Ah thanks - wasn't aware of that functionality. I just reviewed the function and it will work for us.

Reply:
Resolved by using
`dtype=torch.float32 if is_in_onnx_export_mode() else query_layer.dtype`